src/Rules/Functions/PrintfParameterTypeRule.php

```diff
@@ -0,0 +1,145 @@
+use function array_key_exists;
+use function count;
+use function sprintf;
+
+/**
+ * @implements Rule<Node\Expr\FuncCall>
+ */
+final class PrintfParameterTypeRule implements Rule
+{
+
+	private const FORMAT_ARGUMENT_POSITIONS = [
+		'printf' => 0,
+		'sprintf' => 0,
+		'fprintf' => 1,
+	];
+	private const MINIMUM_NUMBER_OF_ARGUMENTS = [
+		'printf' => 1,
+		'sprintf' => 1,
+		'fprintf' => 2,
+	];
+
+	public function __construct(
+		private PrintfHelper $printfHelper,
+		private ReflectionProvider $reflectionProvider,
+		private RuleLevelHelper $ruleLevelHelper,
+	)
+	{
+	}
+
+	public function getNodeType(): string
+	{
+		return Node\Expr\FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!($node->name instanceof Node\Name)) {
+			return [];
+		}
+
+		if (!$this->reflectionProvider->hasFunction($node->name, $scope)) {
+			return [];
+		}
+
+		$functionReflection = $this->reflectionProvider->getFunction($node->name, $scope);
+		$name = $functionReflection->getName();
+		if (!array_key_exists($name, self::FORMAT_ARGUMENT_POSITIONS)) {
+			return [];
+		}
+
+		$formatArgumentPosition = self::FORMAT_ARGUMENT_POSITIONS[$name];
+
+		$args = $node->getArgs();
+		foreach ($args as $arg) {
+			if ($arg->unpack) {
+				return [];
+			}
+		}
+		$argsCount = count($args);
+		if ($argsCount < self::MINIMUM_NUMBER_OF_ARGUMENTS[$name]) {
+			return []; // caught by CallToFunctionParametersRule
+		}
+
+		$formatArgType = $scope->getType($args[$formatArgumentPosition]->value);
+		$formatArgTypeStrings = $formatArgType->getConstantStrings();
+
+		// Let's start simple for now.
+		if (count($formatArgTypeStrings) !== 1) {
+			return [];
+		}
+
+		$formatString = $formatArgTypeStrings[0];
+		$format = $formatString->getValue();
+		$acceptingTypes = $this->printfHelper->getPrintfPlaceholderAcceptingTypes($format);
+		$errors = [];
+		$typeAllowedByCallToFunctionParametersRule = TypeCombinator::union(
+			new StringAlwaysAcceptingObjectWithToStringType(),
+			new IntegerType(),
+			new FloatType(),
+			new BooleanType(),
+			new NullType(),
+		);
+
+		for ($i = $formatArgumentPosition + 1, $j = 0; $i < $argsCount; $i++, $j++) {
+			// Some arguments may be skipped entirely.
+			if (! array_key_exists($j, $acceptingTypes)) {
+				continue;
+			}
+
+			[$acceptingName, $acceptingCb] = $acceptingTypes[$j];
+			$argType = $this->ruleLevelHelper->findTypeToCheck(
+				$scope,
+				$args[$i]->value,
+				'',
+				$acceptingCb,
+			)->getType();
+
+			if ($argType instanceof ErrorType || $acceptingCb($argType)) {
+				continue;
+			}
+
+			// This is already reported by CallToFunctionParametersRule
+			if (
+				!$this->ruleLevelHelper->accepts(
+					$typeAllowedByCallToFunctionParametersRule,
+					$argType,
+					$scope->isDeclareStrictTypes(),
+				)->result
+			) {
+				continue;
+			}
+
+			$errors[] = RuleErrorBuilder::message(
+				sprintf(
+					'Placeholder #%d of function %s expects %s, %s given',
```

Comment:
It'd also be really useful to add the actual placeholder (`%f` for example) in the message here.